### DISCUSSION TITLE
Как работает адаптив?

### DISCUSSION THREAD

**Discussion by scffs** [VKCOM/VKUI#5385] [Q&A] [answered]
Судя по доке в `<AdaptivityProvider>` мы передаем ширину, но как ее определять? Если я передам `ViewWidth.SMALL_MOBILE`, то всегда будет мобильный дизайн, но что делать для Desktop?

Изначально я думал, что `const { viewWidth } = useAdaptivityConditionalRender();` сам определяет ширину, но это оказалось не так.

В общем вопрос такой - как определить ширину экрана юзера, чтобы в зависимости от этого левая колонка скрывалась и появлялся таббар

![image](https://github.com/VKCOM/VKUI/assets/67464545/4e2a115d-64cb-4b28-9f27-7ae7d69f2eb2)

**Comment by BlackySoul** [comment]
Привет! Про адаптивность расписано [вот здесь](https://vkcom.github.io/VKUI/#/Adaptivity) и ещё можно почитать [этот](https://github.com/VKCOM/VKUI/blob/master/docs/ADAPTIVITY_GUIDE.md) гайд 

Ещё есть пример в доке с требующимся функционалом [здесь](https://vkcom.github.io/VKUI/#/Epic) 

Если коротко, то наша адаптивность построена на CSS Media Queries. Хук `useAdaptivityConditionalRender` возвращает параметры адаптивности, на которые нужно опираться при построение интерфейсов. 
Например, нам нужно показывать `Tabbar` только на планшетах и мобильных устройствах. Будем опираться на параметр `viewWidth`.

```
const { viewWidth } = useAdaptivityConditionalRender();

...
// Где-то в коде по условию отрисовываем tabbar
tabbar={
            viewWidth.tabletMinus && (
              <Tabbar className={viewWidth.tabletMinus.className}>
 ...
```

Данный код будет выполняться для всех устройств с экраном меньшим или равным планшету. Если в `AdaptivityProvider` было определено, что `viewWidth=ViewWidth.DESKTOP`, то `viewWidth.tabletMinus` всегда будет возвращать `false`, иначе будет возвращена мета-информация, в частности, объект с `className`, который нужно навесить на компонент, который необходимо показывать/скрывать. Соответственно, при изменении размеров экрана у компонента (`Tabbar` в примере) будет меняться css-свойство `display`, отвечающее за отображение на экране.
Замечание: если в `AdaptivityProvider` ничего не передавать, то все параметры вычисляются автоматически, подстраиваясь под текущие условия.

**Reply by scffs** [reply]
Привет, что-то не очень вычисляется, если честно)

![image](https://github.com/VKCOM/VKUI/assets/67464545/6fdbf123-30e6-461d-ba42-2e475b328794)

![image](https://github.com/VKCOM/VKUI/assets/67464545/f873d27e-349b-44ae-9759-215c2b9498c4)

![image](https://github.com/VKCOM/VKUI/assets/67464545/3ef3568d-6e48-4f9c-9ec1-f296fec5b48c)

**Reply by scffs** [reply]
Если передать `viewWidth={ViewWidth.MOBILE}`, то всё ок, но такой вариант не подходит

![image](https://github.com/VKCOM/VKUI/assets/67464545/b2520c76-4d5d-4036-8d59-51e599a9ae09)

![image](https://github.com/VKCOM/VKUI/assets/67464545/7d02a158-b4ec-44b5-8063-161e9b996f1c)


**UPD**
А если передать в `SplitCol` `className={viewWidth.tabletPlus.className}`  без передачи `viewWidth={ViewWidth.DESKTOP}` в `AdaptivityProvider`, то он не рендерится на больших экранах

![image](https://github.com/VKCOM/VKUI/assets/67464545/dc864ffd-da2c-436e-a4bb-4950cd64b3fa)

**Comment by BlackySoul** [comment]
Извини за такой долгий ответ.

А можно где-то посмотреть всю структуру приложения полностью? Я правильно понимаю, что ты на десктопах пытаешься скрыть `Tabbar`? Тогда почему есть некий  `TabbarDesktop`?) 

У тебя веб-приложение или в контексте мини-аппов? Установлен ли пакет `vk-bridge`?

**Reply by scffs** [reply]
Снова привет, DesktopTabbar - левая колонка на пк, название компонента просто кривое немного))

Весь проект тут, если есть время, то можешь чекнуть
 https://github.com/ExercisesMiniApp/Frontend

Если захочешь запустить проект, то чтобы фронт запустить без бекенда, то ты можешь просто запросы к АПИ удалить, они где-то в FetchUser.ts лежат 

Логика отображения таббара и левой колонки в src/App.tsx

Там сейчас элементам не задаётся className из viewWidth, но у меня с ним вообще ничего не отображается)))

**Reply by BlackySoul** [reply, marked as answer]
Хмм, вроде бы у меня все работает, как ожидается

https://github.com/VKCOM/VKUI/assets/7431217/bf5625b5-4593-453f-a5f9-fb3768cb4791

Изменения такие (ещё добавила в `TabbarMobile` проброс пропа `className`): 

```
const { viewWidth } = useAdaptivityConditionalRender();

  const MobileTabbar = viewWidth.tabletMinus && (
    <Suspense>
      <TabbarMobile className={viewWidth.tabletMinus.className} />
    </Suspense>
  );

  const DesktopTabbar = viewWidth.tabletPlus && (
    <SplitCol
      fixed
      width={280}
      maxWidth={280}
      autoSpaced
      className={viewWidth.tabletPlus.className}
    >
      <Suspense>
        <TabbarDesktop />
      </Suspense>
    </SplitCol>
  );
```

Из того, что заметила - `Tabbar` ещё необходимо вкладывать в `Epic`, по структуре должно быть [как здесь](https://vkcom.github.io/VKUI/#/Epic).



UPD.: увидела, что ты в контексте мини-приложений запускаешь, сейчас попробую тоже
UPD.:  в контексте мини-приложений тоже работает, колонка слева скрыта

**Reply by scffs** [reply]
Интересненько... Позже смогу чекнуть и отпишу, надеюсь, что поможет

**Reply by scffs** [reply]
В браузере действительно всё адаптируется, но в самом mini app у меня таббар, а не `SplitCol`. SplitCol скрывается, если viewWidth < 1024px, но в настройках приложения нельзя поставить ширину > 1000px 

![image](https://github.com/VKCOM/VKUI/assets/67464545/2c4c1810-8141-4c3a-8f66-7f412f2bf95a)

![image](https://github.com/VKCOM/VKUI/assets/67464545/5c511105-b940-4d07-84f1-00147f53c625)

**Reply by BlackySoul** [reply]
А почему в mini app'е должен быть `SplitCol`? Так и должно же работат, т.е. ширина меньше `1024px`, колонка `SplitCol` скрыта

**Reply by scffs** [reply]
Я вообще думал в аппах тоже SplitCol бывает, а то странно как-то, заходишь с ПК, а у тебя дизан как на мобилке.. Ну ок)

**Reply by scffs** [reply]
В таком случае Снекбар скрывается за таббаром...

![image](https://github.com/VKCOM/VKUI/assets/67464545/b2a16e98-0965-4301-9355-8a2b8b3b8d4f)

**Reply by BlackySoul** [reply]
А `Tabbar` завернут в `Epic`, как я выше писала?

**Reply by scffs** [reply]
![image](https://github.com/VKCOM/VKUI/assets/67464545/d43e2cc2-345e-463f-96df-d23c0b9a5d0f)

![image](https://github.com/VKCOM/VKUI/assets/67464545/fdd22d0c-98a4-454a-a346-10ce105bfc8b)

))

**Reply by scffs** [reply]
А, там внутри Epic надо все view/panel тож разместить

**Reply by scffs** [reply]
А, я понял. Всё-таки надо один Epic и туда передавать tabbar, что в моем случае не очень удобно.. Помогло, спасибо, пойду рефакторить теперь)
